File: code/CPP/billion_simulrun.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>
#include <random>
#include <ctime>
#include <chrono>
#include <set>
#include <sstream>
#include <limits>
#include <cstring>
#include <sys/mman.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

class MemoryMappedFile {
private:
    void* data;
    size_t size;
    int fd;

public:
    MemoryMappedFile(const std::string& filename, size_t file_size) {
        fd = open(filename.c_str(), O_RDONLY);
        if (fd == -1) {
            throw std::runtime_error("Failed to open file: " + filename);
        }
        
        size = file_size;
        data = mmap(nullptr, size, PROT_READ, MAP_PRIVATE, fd, 0);
        if (data == MAP_FAILED) {
            close(fd);
            throw std::runtime_error("Failed to mmap file: " + filename);
        }
    }

    ~MemoryMappedFile() {
        if (data != MAP_FAILED) {
            munmap(data, size);
        }
        if (fd != -1) {
            close(fd);
        }
    }

    template<typename T>
    T* get() {
        return static_cast<T*>(data);
    }
};

std::vector<uint32_t> load_shape(const std::string& filename) {
    std::ifstream file(filename, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Failed to open shape file");
    }
    
    // Simple .npy parser for shape (assumes specific format)
    file.seekg(0, std::ios::end);
    size_t file_size = file.tellg();
    file.seekg(0, std::ios::beg);
    
    // Skip numpy header (typically 128 bytes for simple arrays)
    char header[256];
    file.read(header, 128);
    
    // Read shape values (this is simplified - full .npy parsing is more complex)
    std::vector<uint32_t> shape;
    // For this example, we'll assume you know the shape or parse it differently
    // You might want to use a proper .npy library or store shape in a simpler format
    
    return shape;
}

std::set<int> load_completed(const std::string& filename) {
    std::set<int> completed;
    std::ifstream file(filename);
    std::string line;
    
    while (std::getline(file, line)) {
        if (!line.empty()) {
            completed.insert(std::stoi(line));
        }
    }
    
    return completed;
}

void compute_distances(const int8_t* X, int n_vectors, int dim, 
                       int source, float* dist_from_source) {
    // Compute: -2 * (X @ X[source]) + (X[source] @ X[source])
    const int8_t* source_vec = X + (size_t)source * dim;
    
    // Compute X[source] @ X[source]
    float source_norm = 0.0f;
    for (int d = 0; d < dim; d++) {
        source_norm += source_vec[d] * source_vec[d];
    }
    
    // Compute for all vectors
    #pragma omp parallel for
    for (int i = 0; i < n_vectors; i++) {
        const int8_t* vec = X + (size_t)i * dim;
        float dot_product = 0.0f;
        
        for (int d = 0; d < dim; d++) {
            dot_product += vec[d] * source_vec[d];
        }
        
        dist_from_source[i] = -2.0f * dot_product + source_norm;
    }
}

int main() {
    const std::string DATASET = "spacev1b";
    const std::string METRIC = "euclidean";
    const std::string SAVEPATH = "/scratch/pa2439/Projects/ANN-Search/navigable_graph_results/results";
    const std::string BINARY_FILE = "/scratch/pa2439/Projects/ANN-Search/datasets/SPACEV1B/vectors_int.mmap";
    const std::string SQ_NORMS_FILE = "/scratch/pa2439/Projects/ANN-Search/datasets/SPACEV1B/sq_vector_norms.mmap";
    const std::string SHAPE_FILE = "/scratch/pa2439/Projects/ANN-Search/datasets/SPACEV1B/vectors_shape.npy";
    
    std::cout << "Building graph on " << DATASET << std::endl;
    
    // For simplicity, hardcode shape or read from a simpler format
    // You'll need to either parse .npy properly or save shape in text format
    uint32_t n_vectors = 1000000000; // 1B vectors - adjust as needed
    uint32_t dim = 100; // dimension - adjust as needed
    
    // Memory map the vectors file
    size_t vectors_size = (size_t)n_vectors * dim * sizeof(int8_t);
    MemoryMappedFile vectors_mmap(BINARY_FILE, vectors_size);
    int8_t* X = vectors_mmap.get<int8_t>();
    
    std::cout << "Loaded vectors: " << n_vectors << " x " << dim << std::endl;
    
    // Memory map squared norms
    size_t norms_size = n_vectors * sizeof(float);
    MemoryMappedFile norms_mmap(SQ_NORMS_FILE, norms_size);
    float* sq_norms = norms_mmap.get<float>();
    
    std::cout << "Loaded squared norms" << std::endl;
    
    // Load completed sources
    std::set<int> completed = load_completed(SAVEPATH + "/" + DATASET + "-" + METRIC + "-computed.txt");
    
    // Create shuffled array of sources
    std::vector<uint32_t> all_sources(n_vectors);
    for (uint32_t i = 0; i < n_vectors; i++) {
        all_sources[i] = i;
    }
    
    std::random_device rd;
    std::mt19937 gen(rd());
    std::shuffle(all_sources.begin(), all_sources.end(), gen);
    
    // Process only first source (as in Python [:1])
    for (size_t idx = 0; idx < 1 && idx < all_sources.size(); idx++) {
        uint32_t source = all_sources[idx];
        
        if (completed.find(source) != completed.end()) {
            continue;
        }
        
        // Compute neighborhood of source
        std::vector<uint32_t> edges;
        edges.push_back(source);
        
        std::vector<uint8_t> active(n_vectors, 1);
        active[source] = 0;
        
        std::cout << "Computing pairwise distances from " << source << std::endl;
        
        auto start = std::chrono::high_resolution_clock::now();
        
        std::vector<float> dist_from_source(n_vectors);
        compute_distances(X, n_vectors, dim, source, dist_from_source.data());
        
        auto end = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double> elapsed = end - start;
        std::cout << "Took " << elapsed.count() << " seconds to compute pairwise distances" << std::endl;
        
        start = std::chrono::high_resolution_clock::now();
        
        // Count active nodes
        size_t active_count = n_vectors - 1;
        
        while (active_count > 0) {
            std::cout << "Degree: " << edges.size() - 1 << ", Left: " << active_count << std::endl;
            
            // Find minimum distance among active nodes
            float min_dist = std::numeric_limits<float>::infinity();
            uint32_t waypoint = 0;
            
            for (uint32_t i = 0; i < n_vectors; i++) {
                if (active[i] && dist_from_source[i] < min_dist) {
                    min_dist = dist_from_source[i];
                    waypoint = i;
                }
            }
            
            active[waypoint] = 0;
            active_count--;
            edges.push_back(waypoint);
            
            // Compute distances from waypoint and prune
            std::vector<float> dist_from_waypoint(n_vectors);
            compute_distances(X, n_vectors, dim, waypoint, dist_from_waypoint.data());
            
            for (uint32_t i = 0; i < n_vectors; i++) {
                if (active[i] && dist_from_waypoint[i] < dist_from_source[i]) {
                    active[i] = 0;
                    active_count--;
                }
            }
        }
        
        end = std::chrono::high_resolution_clock::now();
        elapsed = end - start;
        std::cout << "Took " << elapsed.count() << " seconds to compute neighborhood" << std::endl;
        
        // Write adjacency list
        std::ofstream adj_file(SAVEPATH + "/adj-list-" + DATASET + "-" + METRIC + ".txt", 
                               std::ios::app);
        for (size_t i = 0; i < edges.size(); i++) {
            adj_file << edges[i];
            if (i < edges.size() - 1) adj_file << ",";
        }
        adj_file << "\n";
        adj_file.close();
        
        // Write completed
        std::ofstream comp_file(SAVEPATH + "/" + DATASET + "-" + METRIC + "-computed.txt",
                                std::ios::app);
        comp_file << source << "\n";
        comp_file.close();
    }
    
    std::cout << "Done with " << DATASET << std::endl;
    
    return 0;
}

File: code/billion_simulrun.py
import numpy as np
import argparse
from tqdm import tqdm
from time import time
import psutil

def main():
    DATASET = "spacev1b" # Get dataset name from argument
    METRIC = 'euclidean'

    print("Building graph on", DATASET, flush=True)
    SAVEPATH = "/scratch/pa2439/Projects/ANN-Search/navigable_graph_results/results"
    # BINARY_FILE = "/scratch/pa2439/Projects/ANN-Search/datasets/SPACEV1B/vectors.mmap"
    BINARY_FILE = "/scratch/pa2439/Projects/ANN-Search/datasets/SPACEV1B/vectors_int.mmap"
    SQ_NORMS_FILE = "/scratch/pa2439/Projects/ANN-Search/datasets/SPACEV1B/sq_vector_norms.mmap"
    dataset_shape = np.load("/scratch/pa2439/Projects/ANN-Search/datasets/SPACEV1B/vectors_shape.npy")
    MEMPOINT =  10 ** 8

    # X = np.memmap(BINARY_FILE, mode='r', dtype='float32', shape=dataset_shape)
    X = np.memmap(BINARY_FILE, mode='r', dtype='int8', shape=dataset_shape)
    print("Loaded vectors:", X.shape, flush=True)

    # sq_norms = np.memmap(SQ_NORMS_FILE, mode='r', dtype='float32', shape=X.shape[0])
    # print("Loaded squared norms", flush=True)

    metric = 'euclidean'

    # assume no other program is trying to edges simultaneously
    with open(f"{SAVEPATH}/{DATASET}-{metric}-computed.txt", 'a+') as f:
        f.seek(0)
        completed = set([int(line.strip()) for line in f if line.strip()])
    
    all_sources = np.arange(X.shape[0], dtype=np.uint32)
    np.random.shuffle(all_sources)

    for source in all_sources[:1]:
        if source not in completed:
            
            ### compute neighborhood of source ###

            edges = [source]

            active = np.zeros(X.shape[0], dtype=np.uint8)
            active[source] = 0

            
            print(f"Computing pairwise distances from {source}", flush=True)
            start = time()
            dist_from_source = -2 * (X @ X[source]) + (X[source] @ X[source])
            # dist_from_source = -2 * (X @ X[source]) + sq_norms[source]
            print(f"Took {time() - start} seconds to compute pairwise distances", flush=True)

            start = time()
            while np.sum(active) > MEMPOINT:
                matvectime = time()
                masked_dist = np.where(active > 0, dist_from_source, np.inf)
                waypoint = np.argmin(masked_dist)
                active[waypoint] = 0
                edges.append(waypoint)

                dist_from_waypoint = -2 * (X @ X[waypoint]) + (X[waypoint] @ X[waypoint])
                active[dist_from_waypoint < dist_from_source] = 0
                print(f"Degree: {len(edges) - 1}, Left: {np.sum(active)}, Took: {time() - matvectime}", flush=True)
            
            # active_indices = active[active > 0]
            # X_inmem = X[active_indices]
            # dist_inmem = dist_from_source[active_indices]
            # active_inmem = active[active_indices]
            # print(f"moving into memory..., {len(active_indices)}, {active_inmem.shape}, {X_inmem.shape}, {active_inmem[0]}", flush=True)

            # while np.sum(active_inmem) > 0:
            #     matvectime = time()
            #     masked_dist = np.where(active_inmem > 0, dist_inmem, np.inf)
            #     waypoint = np.argmin(masked_dist)
            #     active_inmem[waypoint] = 0
            #     edges.append(waypoint)

            #     dist_from_waypoint = -2 * (X_inmem @ X_inmem[waypoint]) + (X_inmem[waypoint] @ X_inmem[waypoint])
            #     active_inmem[dist_from_waypoint < dist_inmem] = 0
            #     print(f"Degree: {len(edges) - 1}, Left: {np.sum(active_inmem)}, Took: {time() - matvectime}", flush=True)

            ######################################

            print(f"Took {time()-start} seconds to compute neighborhood", flush=True)

            with open(f"{SAVEPATH}/adj-list-{DATASET}-{metric}.txt", 'a') as adj:
                # FIX: Added closing parenthesis and newline character
                adj.write(','.join([str(e) for e in edges]) + '\n')
            
            with open(f"{SAVEPATH}/{DATASET}-{metric}-computed.txt", 'a') as f:
                f.write(f"{source}\n")
        
    print(f"Done with {DATASET}")


def compute_sq_norms(X_memmap, memory_factor=0.8):
    """Computes the squared L2-norm for every vector in a large memmap."""
    
    N, D = X_memmap.shape
    
    # Get available RAM and determine memory limit for the chunk
    available_ram_bytes = psutil.virtual_memory().available
    chunk_ram_limit = available_ram_bytes * memory_factor
    
    # Calculate adaptive chunk size
    bytes_per_vector = D * (X_memmap.dtype.itemsize + np.dtype(np.float32).itemsize)
    max_chunk_size = int(chunk_ram_limit / bytes_per_vector)
    chunk_size = max(1, min(max_chunk_size, N))

    print(f"Total available RAM: {available_ram_bytes / 1024**3:.2f} GB")
    print(f"Chunk limit ({memory_factor*100:.0f}%): {chunk_ram_limit / 1024**3:.2f} GB")
    print(f"Using chunk size: {chunk_size} vectors")

    # Initialize result array
    sq_norms = np.empty(N, dtype=np.int16)

    # Process data in chunks
    for start in range(0, N, chunk_size):
        end = min(start + chunk_size, N)

        X_chunk = X_memmap[start:end] 
        
        # Vectorized squaring and summation
        sq_norms_chunk = np.sum(X_chunk * X_chunk, axis=1, dtype=np.int16)
        
        sq_norms[start:end] = sq_norms_chunk

    return sq_norms

def compute_inner_products(X_memmap, query_vector, memory_factor=0.8):
    """Computes the inner product of a query vector against every vector in a large memmap."""
    
    N, D = X_memmap.shape
    
    if query_vector.shape != (D,):
        raise ValueError(f"Query vector must be {D}-dimensional, but got {query_vector.shape}")
    
    q = query_vector.astype(np.float32)

    # Determine Memory Limit
    available_ram_bytes = psutil.virtual_memory().available
    chunk_ram_limit = available_ram_bytes * memory_factor
    
    # Calculate adaptive chunk size
    bytes_per_vector = D * (X_memmap.dtype.itemsize + np.dtype(np.float32).itemsize)
    max_chunk_size = int(chunk_ram_limit / bytes_per_vector)
    chunk_size = max(1, min(max_chunk_size, N))

    print(f"Total available RAM: {available_ram_bytes / 1024**3:.2f} GB", flush=True)
    print(f"Chunk limit ({memory_factor*100:.0f}%): {chunk_ram_limit / 1024**3:.2f} GB", flush=True)
    print(f"Using chunk size: {chunk_size} vectors", flush=True)

    # Initialize result array
    inner_products = np.empty(N, dtype=np.float32)

    # Process data in chunks
    for start in tqdm(range(0, N, chunk_size)):
        end = min(start + chunk_size, N)

        X_chunk = X_memmap[start:end] 
        
        # Convert to float32 to enable BLAS-optimized matrix multiplication
        X_float = X_chunk.astype(np.float32)

        # Perform the Matrix-Vector (GEMV) multiplication: X_chunk * q
        products_chunk = X_float @ q
        
        inner_products[start:end] = products_chunk
    
    return inner_products

if __name__ == "__main__":
    main()
